core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -77,8 +77,9 @@ class LogSegment private[log] (val log: FileRecords,
     timeIndex.resize(size)
   }
 
-  def sanityCheck(timeIndexFileNewlyCreated: Boolean): Unit = {
-    if (lazyOffsetIndex.file.exists) {
+  def sanityCheck(timeIndexFileNewlyCreated: Boolean, isActiveSegment: Boolean): Unit = {
+    // We allow for absence of offset index file only for an empty active segment.
+    if ((isActiveSegment && size == 0) || lazyOffsetIndex.file.exists) {
```

Comment:
I am wondering why the active segment will be missing the offset index file during a clean shutdown. When we load the segments during broker restart, we call resizeIndexes() on the last segment. This should trigger the creation of the offset index file, which will be flushed on broker shutdown.

Reply:
> When we load the segments during broker restart, we call resizeIndexes() on the last segment. This should trigger the creation of the offset index file, which will be flushed on broker shutdown.

The sanityCheck is called before resizeIndexes.

It appears you are talking about we start the broker then immediately shut it down. In between the active segment may have been changed, and if the new one is empty, no index file is created.

Reply:
I am still trying to understand if the missing index is the result of a clean shutdown or a hard shutdown. When will roll a segment, the index on the new active segment is created lazily. However, during a clean shutdown, we force flush the active segment, which should trigger the creation of an empty index file because the following method is used in segment flush.

`  def offsetIndex: OffsetIndex = lazyOffsetIndex.get
`
On a hard shutdown, it's possible for the offset index to be missing. However, in that case, the offset index can be missing even when the log is not empty. So, I am wondering how common of an issue that we are fixing.

Reply:
@junrao: When the `UnifiedLog` is [flushed during clean shutdown](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/LogManager.scala#L527), we flush the `LocaLog` [until the logEndOffset](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/UnifiedLog.scala#L1508). Here an empty active segment is not included in the list of candidate segments to be flushed. The reason is that during `LocalLog.flush()`, the `LogSegments.values(recoveryPoint, logEndOffset)` call [here](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/LocalLog.scala#L171) does not select the empty active segment ([doc](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/LogSegments.scala#L127-L131)), because, the logEndOffset would match the base offset of the empty active segment and thus get ommitted. So, prior to clean shutdown if the empty active segment's offset index was never created before, then, the offset index will not be created during clean shutdown because the empty active segment is never flushed.

The above is shown in the following passing unit test:

```
@Test
def testFlushEmptyActiveSegmentDoesNotCreateOffsetIndex(): Unit = {
    // Create an empty log.
    val logConfig = LogTestUtils.createLogConfig(segmentBytes = 1024 * 1024)
    val log = createLog(logDir, logConfig)
    val oneRecord = TestUtils.records(List(
      new SimpleRecord(mockTime.milliseconds, "a".getBytes, "value".getBytes)
    ))

    // Append a record and flush. Verify that there exists only 1 segment.
    log.appendAsLeader(oneRecord, leaderEpoch = 0)
    assertEquals(1, log.logEndOffset)
    log.flush()
    assertEquals(1, log.logSegments.size)
    assertTrue(UnifiedLog.logFile(logDir, 0).exists())
    assertTrue(UnifiedLog.offsetIndexFile(logDir, 0).exists())
    assertFalse(UnifiedLog.logFile(logDir, 1).exists())
    assertFalse(UnifiedLog.offsetIndexFile(logDir, 1).exists())

    // Roll the log and verify that the new active segment's offset index is missing.
    log.roll()
    assertEquals(2, log.logSegments.size)
    assertTrue(UnifiedLog.logFile(logDir, 0).exists())
    assertTrue(UnifiedLog.offsetIndexFile(logDir, 0).exists())
    assertTrue(UnifiedLog.logFile(logDir, 1).exists())
    assertFalse(UnifiedLog.offsetIndexFile(logDir, 1).exists())

    // Flush the log and once again verify that the active segment's offset index is still missing.
    log.flush()
    assertTrue(UnifiedLog.logFile(logDir, 0).exists())
    assertTrue(UnifiedLog.offsetIndexFile(logDir, 0).exists())
    assertTrue(UnifiedLog.logFile(logDir, 1).exists())
    assertFalse(UnifiedLog.offsetIndexFile(logDir, 1).exists())

    // Close the log and verify that the active segment's offset index is still missing.
    log.close()
    assertTrue(UnifiedLog.logFile(logDir, 0).exists())
    assertTrue(UnifiedLog.offsetIndexFile(logDir, 0).exists())
    assertTrue(UnifiedLog.logFile(logDir, 1).exists())
    assertFalse(UnifiedLog.offsetIndexFile(logDir, 1).exists())
}
```

This PR mainly fixes a logging issue in the code. For example, one situation where the issue happens more frequently is the following: Imagine there exists a topic with very low ingress traffic in some/all partitions. Imagine that for this topic the retention setting causes all existing segments to expire and get removed. In such a case, [we roll the log to create an active segment](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/UnifiedLog.scala#L1352-L1354). This ensures there is at least one segment remaining in the `LocalLog` when the retention loop completes. However we don't create the offset index for the active segment until the first append operation. Now before the first append, if the Kafka cluster is rolled then we will see [this false negative corruption error message during recovery.](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/LogLoader.scala#L336-L337)

This PR fixes the logging problem by ignoring the absence of offset index for an empty active segment during recovery.

Reply:
@kowshik : Thanks for the great explanation. It makes sense to me now.

If we don't flush the only empty log segment during clean shutdown, we could lose the log segment file as well, which causes the replica to lose track of logEndOffset. I am wondering if we should force flush the empty log segment during clean shutdown too.

Reply:
@junrao That is a really good point. Yes, it looks useful to force flush the empty segment during clean shutdown. I couldn't find a good reason that we don't do it today, other than the fact that the flush call [here](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/UnifiedLog.scala#L1508) probably did not cover for the case where the active segment could be empty. I'm thinking a simple fix could be the below change to `UnifiedLog`:

```
  /**
   * Flush all local log segments, including the active segment.
   */
  def flush(): Unit = flush(logEndOffset + 1)
```

Could I open a separate PR with this improvement?

Reply:
@kowshik : If we flush the active segment, do we still need the changes in sanityCheck()?

Reply:
We only want to flush empty active segments during shutdown right?

Reply:
When shutting down, we want to flush all segments, including the empty active segment.

Reply:
@junrao It's a really good question. https://github.com/apache/kafka/pull/8346 is the PR where lazy indexes were introduced. The description says it eliminates unnecessary disk accesses and memory map operations while
deleting, renaming or closing offset and time indexes. As a result, a significant shutdown time improvement was noticed. Supposing if we flushed the empty active segment, could the shutdown time regress?

Reply:
Will `def flush(): Unit = flush(logEndOffset + 1)` trigger flushing empty active segments every time we roll a segment, not only during shutdown?

Reply:
@ccding When we roll a segment, we explicitly ensure to flush only the old segment. See [this LOC](https://github.com/apache/kafka/blob/c6aeb5c5546f34276879c1eef4199d208f1c23dc/core/src/main/scala/kafka/log/UnifiedLog.scala#L1501).

Reply:
@kowshik : #8346 tries to avoid opening the index during close() when the index is not opened yet. This applies to existing segments on broker restart. For active segment, we typically need to open the index anyway. So, we probably don't need to optimize the rare case when it's empty. Plus, the danger of losing logEndOffset is a bigger concern than avoiding the cost of opening one index file.

Reply:
@junrao Sounds good to me. We can flush the active segment during clean shutdown. That's a very elegant way to handle this problem.

@ccding Would you like to update the PR with the approach proposed by @junrao?

Reply:
Updated. Please let me know if I misunderstood anything.